UpSample GPU Porting

xmnlab opened the pull request:
resolves #16158

xmnlab committed:
Porting in progress

rgommers committed:
Fix CUDA error checks in UpSampleBicubic2d.cu

This will now give more informative errors:

  RuntimeError: CUDA error: too many resources requested for launch

instead of

  RuntimeError: Failed with error code 0

rgommers committed:
Set launch bounds to 1024 for upsamplebicibic2d

See gh-8103 for other reports of this issue.

rgommers commented:
Fewer errors than before, 4 instead of 8 with the same message:
```
RuntimeError: CUDA error: too many resources requested for launch
```
This seems to be an existing problem: gh-8103.

@skrah could you have a look at this and tell us what you think?

skrah commented:
Looking at it very briefly, the `Jetson TX` referenced in the issue only has 256 cores.

So while the issue looks the same, I'd probably not expect it on the CI platforms.  FWIW, some recent CI tests in other issues are green.

skrah commented:
@pytorchbot retest this please.

rgommers commented:
same failures.

skrah commented:
Is this rebased on the latest master (you can also ask the bot to rebase)?

xmnlab commented:
I rebased that yesterday ... also needed to resolve conflicts because 7 days ago upsample files were changed.

xmnlab commented:
@skrah do you have any idea about what could be this problem? or a way to debug that?
also it seems some jobs is taking a lot of time to build, for example, for some job the estimate time is 19h .. not sure if it the regular estimate ..

skrah commented:
@xmnlab If you can't reproduce it at home it seems hard to debug other than reading at the diffs again.  I can take a look on Monday, it's getting a bit late here.

skrah commented:
Also, has anyone found a way to show the actual hardware used on the CI in detail?

xmnlab commented:
@skrah I am working in parallel on a paperspace environment .. it tooks a lot of time it is running with 8 cpu cores.

skrah commented:
On Fri, May 03, 2019 at 07:12:35PM +0000, Ivan Ogasawara wrote:
> @skrah I am working in parallel on a paperspace environment .. it tooks a lot of time it is running with 8 cpu cores.

But have you ever been able to reproduce this issue on paperspace?

rgommers commented:
> But have you ever been able to reproduce this issue on paperspace?

I can reproduce it locally: Arch Linux, CUDA 10.0, RTX2070 GPU. Given the CI failures, I think it should be reproducible for multiple CUDA and GPU versions. I just haven't worked on any CUDA code before, and am short on time, so I'd rather not dig too deep.

xmnlab commented:
not yet .. my last building was with a my previous commit ... I will work again in this task in some minutes :)

rgommers commented:
@xmnlab your previous commit had the same failure though (except for the less clear exception message), and it seems quite reproducible. So I think you'll see it now.

skrah commented:
On Fri, May 03, 2019 at 12:19:42PM -0700, Ralf Gommers wrote:
> > But have you ever been able to reproduce this issue on paperspace?
> 
> I can reproduce it locally: Arch Linux, CUDA 10.0, RTX2070 GPU. Given the CI failures, I think it should be reproducible for multiple CUDA and GPU versions. I just haven't worked on any CUDA code before, and am short on time, so I'd rather not dig too deep.

Ah, thanks.  @xmnlab, then I guess just compiling with "DEBUG=1" and stepping through the offending code with gdb may be the fastest way.

xmnlab commented:
@skrah thanks! I will try that!

xmnlab committed:
Fixed register issue, recommendation from @skrah
Fixing launch bounds
Move back from int64_t to int
Changed at::zero to at::empty_like
Use cuda::ATenCeilDiv, removed unncessary += op
Decreasing max threads per block
Removing declaration on THNN

xmnlab commented:
it seems just UpSampleBicubic2d is using upsample_get_value_bounded (https://github.com/pytorch/pytorch/pull/19630/files#diff-5092da792c30694ee4adf0d0ae2a37c6R171) and upsample_increment_value_bounded (https://github.com/pytorch/pytorch/pull/19630/files#diff-5092da792c30694ee4adf0d0ae2a37c6R191)

maybe the problem could be inside one of these functions ... maybe related to the order of indexes (x, y) ...

but the problem seems to be related to cuda block/threads ... so not sure if it is really related to these functions.

skrah commented:
The new code uses far more registers than the existing one. I verified that the both versions actually call the offending test case with `1024` in `blockDim`.  So it's very likely a register issue.

`Existing` uses `64` registers, which seems to be optimal or my card:

```
ptxas info    : 77696 bytes gmem, 72 bytes cmem[3]
ptxas info    : Compiling entry function '_Z23bicubic_interp2d_kerneliddb15THCDeviceTensorIdLi4Ei16DefaultPtrTraitsES1_' for 'sm_61'
ptxas info    : Function properties for _Z23bicubic_interp2d_kerneliddb15THCDeviceTensorIdLi4Ei16DefaultPtrTraitsES1_
    0 bytes stack frame, 0 bytes spill stores, 0 bytes spill loads
ptxas info    : Used 64 registers, 432 bytes cmem[0], 24 bytes cmem[2]
```

`New` uses `124` registers, which is too much for my card:

```
ptxas info    : 77696 bytes gmem, 72 bytes cmem[3]
ptxas info    : Compiling entry function '_ZN2at6native76_GLOBAL__N__52_tmpxft_0000626d_00000000_6_UpSampleBicubic2d_cpp1_ii_b4c1e1f328upsample_bicubic2d_out_frameElddbNS_20PackedTensorAccessorIdLm4ENS_16DefaultPtrTraitsElEES4_' for 'sm_61'
ptxas info    : Function properties for _ZN2at6native76_GLOBAL__N__52_tmpxft_0000626d_00000000_6_UpSampleBicubic2d_cpp1_ii_b4c1e1f328upsample_bicubic2d_out_frameElddbNS_20PackedTensorAccessorIdLm4ENS_16DefaultPtrTraitsElEES4_
    0 bytes stack frame, 0 bytes spill stores, 0 bytes spill loads
ptxas info    : Used 124 registers, 496 bytes cmem[0], 24 bytes cmem[2]
```

With `__launch_bounds__(1024)`, the code again uses `64` registers.

If you use `C10_LAUNCH_BOUNDS_1(1024)` **for both kernels**, the tests pass here.


Now *why* is the regcount higher in the new code? It could be `PackedTensorAccessor`, it could be the fact that many instances of `int` have been changed to `int64_t`. :)

You could experiment or just use the launch bounds.  Other code in `native/cuda` seems to use lower bounds for `blockDim`, too.  1024 seems to be an outlier.

xmnlab commented:
@skrah 

it seems it worked locally! thank you so much!  I really appreciate that!

xmnlab commented:
thanks @rgommers and @skrah for all the help!

@ezyang it is done for review!

ezyang commented:
> Also, has anyone found a way to show the actual hardware used on the CI in detail?

CircleCI runs on AWS, and the I'm fairly sure we've got one of the g3 types, either g3.4xlarge or g3.8xlarge

ezyang commented:
cc @ngimel, if you want to take a look at this (I'm planning to do a review too)

ezyang commented on aten/src/ATen/native/cuda/UpSample.cuh:
```diff
@@ -0,0 +1,254 @@
+/* TODO: move this to a common place */
+template <typename scalar_t>
+__device__ inline scalar_t min(scalar_t a, scalar_t b) {
+  return ((a) < (b)) ? (a) : (b);
```
nit (you don't have to fix this, but FYI), you don't need the extra parentheses if you aren't writing  a macro.

xmnlab replied:
thanks @ezyang for the explanation!

ezyang commented on aten/src/ATen/native/cuda/UpSample.cuh:
```diff
@@ -0,0 +1,254 @@
+  return ((a) > (b)) ? (a) : (b);
+}
+
+static inline void upsample_1d_shape_check(
```
Not from upsampling.h, it's from the respective cu files.

ezyang commented on aten/src/ATen/native/cuda/UpSample.cuh:
```diff
@@ -0,0 +1,254 @@
+template <typename accscalar_t>
+__host__ __forceinline__ static accscalar_t area_pixel_compute_scale(
+    int64_t input_size,
+    int64_t output_size,
```
When you were originally trying to debug the resources problem, @skrah suggest one possible explanation was the use of int64_t where int was previously used. Although this ended up not being the root cause of the resources problem, it is still a legitimate issue to be aware of when writing CUDA kernels: int64 indexing is *noticeably* slower than int indexing, and so you shouldn't just willy nilly convert things to int64 in CUDA land; you may need to produce two kernels (one for int and another for int64). In this case, I think this is unlikely to have caused a problem, since there isn't any indexing involved. The best way to be sure, however, is to do a benchmark of the kernel before and after and ensure it hasn't gotten materially slower. That would be quite helpful here.

ezyang commented on aten/src/ATen/native/cuda/UpSample.cuh:
```diff
@@ -0,0 +1,254 @@
+/* TODO: maybe change channel and batch order in the arguments */
+template <typename scalar_t>
+__device__ __forceinline__ static scalar_t upsample_get_value_bounded(
+    const PackedTensorAccessor<scalar_t, 4> data,
```
I don't think you want to do this. A PackedTensorAccessor is a big struct which contains all of the information necessary so that it can be transmitted across a CUDA kernel call. You may doing an unnecessary copy of all that information with a function signature as done here. At the point you're calling this helper function, you should have decayed the PackedTensorAccessor into a TensorAccessor.

ezyang replied:
More concretely, the suggestion is that this should just take a const reference to a TensorAccessor (or, if you are just looking for the minimal change, I suppose const reference to PackedTensorAccessor would be fine.)

ezyang replied:
Actually, I think this would exactly explain why you were using so many registers. @skrah would you agree with me?

ezyang replied:
There are probably some wiki updates in order here.

skrah replied:
@ezyang Yes, I just checked and using references to `PackedTensorAccessor` gets down the register count from `124` to `80`.

This is when you do:

```
typedef double scalar_t;
typedef double accscalar_t;
```

Still more than the original, but much better.

ezyang replied:
I think the rest of the difference can be explained by the fact that `THCDeviceTensor` defaults to 32-bit indexing where as `PackedTensorAccessor` defaults to 64-bit indexing. I think for the ports, I'd prefer it if we kept the old indexing integer size, so as to preserve both performance characteristics and bugs. It's probably worth working out if upsample works with larger than 8G tensors; it probably doesn't, much like the rest of the code in THC. cc @colesbury 

This also needs a wiki update.

ezyang commented on aten/src/ATen/native/cuda/UpSample.cuh:
```diff
@@ -0,0 +1,254 @@
+/* just affect UpSampleBicubic2d.cu */
+/* TODO: change width and height order in the arguments */
+/* TODO: maybe change x and y order in the arguments */
+/* TODO: maybe change channel and batch order in the arguments */
```
Wow, this is yucky. You can do these in another PR but it would definitely be nice to fix these TODOs.

xmnlab replied:
sounds good @ezyang ! thanks!

ezyang commented on aten/src/ATen/native/cuda/UpSample.cuh:
```diff
@@ -0,0 +1,254 @@
+/* TODO: maybe change channel and batch order in the arguments */
+template <typename scalar_t, typename accscalar_t>
+__device__ __forceinline__ static void upsample_increment_value_bounded(
+    PackedTensorAccessor<scalar_t, 4> data,
```
const ref this

xmnlab replied:
It seems here it would be just ref, as it is changed by the atomicAdd

ezyang commented on aten/src/ATen/native/cuda/UpSample.cuh:
```diff
@@ -0,0 +1,254 @@
+  int64_t access_y =
+      max<int64_t>(min<int64_t>(y, height - 1), static_cast<int64_t>(0));
+  atomicAdd(
+      &data[batch][channel][access_y][access_x], static_cast<scalar_t>(value));
```
Can we add ScalarConvert porting instructions to the wiki please? :) (or, I guess, if they're in your tool, that works too)

ezyang replied:
Yeah, a TODO note SGTM.

xmnlab replied:
thanks @ngimel @ezyang ! I am adding a note there.

ezyang commented on aten/src/ATen/native/cuda/UpSampleBicubic2d.cu:
```diff
@@ -1,25 +1,335 @@
+namespace {
+
+template <typename scalar_t, typename accscalar_t>
+C10_LAUNCH_BOUNDS_1(1024)
```
I've approved many of the ROCm PRs that added (possibly unjustified) launch bounds, because in the interest of expediency it's helpful to hardcode assumptions about the AMD hardware in question, but I'm not sure this launch bound is justified on CUDA. If we look at the launch site:

```
  const int64_t max_threads =
      at::cuda::getCurrentDeviceProperties()->maxThreadsPerBlock;

        upsample_bicubic2d_out_frame<scalar_t, accscalar_t>
            <<<(num_output_elements + max_threads - 1) / max_threads,
               max_threads,
               0,
               stream>>>
```

Although it is difficult to find a device for which `maxThreadsPerBlock > 1024`, to be strictly correct, you should clamp the kernel launch to 1024, if you want to keep these bounds.

My preferred series of resolutions:
1. Check if `const&` on `PackedTensorAccessor` solves the register problem
2. If it doesn't, fix the launch code to clamp to 1024

ezyang replied:
According to @skrah, fixing the PackedTensorAccessor problem solves the register problem. So I would prefer to remove the launch bounds entirely.

ngimel replied:
I haven't looked in detail, but even if it's ok w/o launch bounds on bigger cards, removing launch_bounds is probably going to break jetson (see multpile issues where someone is reporting code breaking on jetson, and the fix is to add launch_bounds), and also future compiler versions can become more register hungry (it happened before).

ezyang replied:
Thank you; @xmnlab, you should listen to Natalia about this. In that case we need to fix the kernel launch.

xmnlab replied:
sounds good, thanks @ngimel @ezyang !

ailzhang reviewed (commented):
Sorry, another request here is we should change the files names to be more general as they're not only doing upsamples but also downsampling as well.

xmnlab commented:
@ailzhang could it be addressed in another PR?

ailzhang commented:
@xmnlab Sure, it's definitely better to rename both cpu & gpu impls together in a separate PR. Just want to make sure we have the correct one after porting.

ngimel commented on aten/src/ATen/native/cuda/UpSampleBilinear2d.cu:
```diff
@@ -1,25 +1,312 @@
+      for (int n = 0; n < batchsize; n++) {
+        for (int c = 0; c < channels; ++c) {
+          const scalar_t val = odata[n][c][h1][w1];
+          idata[n][c][h2][w2] += val;
```
What's the contract here (and in all other *backward_out_* functions)? Is it accumulated into provided grad_input tensor, or filling grad_input with the result of interpolate_backward? Usually it's the latter for *_out functions, so `+=` is wrong?

ngimel commented on aten/src/ATen/native/cuda/UpSampleBicubic2d.cu:
```diff
@@ -28,18 +338,21 @@ Tensor& upsample_bicubic2d_backward_out_cuda(
     bool align_corners) {
-    return at::legacy::th::_thnn_upsample_bicubic2d_backward(
-        grad_output, output_size, input_size, align_corners);
+  Tensor grad_input = at::zeros(input_size, grad_output.options());
```
grad_input is set to zero in the template, no need to do it here.

xmnlab replied:
ok thanks @ngimel ! I changed `at::zeros` to `empty_like`. let me know if you have another suggestion for it.

ngimel commented on aten/src/ATen/native/cuda/UpSampleBilinear2d.cu:
```diff
@@ -1,25 +1,312 @@
+  }
+}
+
+static void upsample_bilinear2d_out_cuda_template(
```
Is this upsample_bicubic2d_out_cuda_template almost verbatim, just launching the different kernel? Can they (and similar cases) be deduplicated?

xmnlab replied:
not sure .. maybe could be .. I am just following the examples to performe the porting. If you have any other example related your thought, maybe we could address that in a follow PR.

ezyang commented on aten/src/ATen/native/cuda/UpSampleBilinear2d.cu:
```diff
@@ -1,25 +1,312 @@
+    //
+    const accscalar_t h1r = area_pixel_compute_source_index<accscalar_t>(
+        rheight, h2, align_corners, /*cubic=*/false);
+    const int64_t h1 = h1r;
```
Use of `int64_t` here is much less benign from a performance perspective, as you're going to index with it.

ezyang commented on aten/src/ATen/native/cuda/UpSampleBilinear2d.cu:
```diff
@@ -1,25 +1,312 @@
+            h1lambda *
+                (w0lambda * idata[n][c][h1 + h1p][w1] +
+                 w1lambda * idata[n][c][h1 + h1p][w1 + w1p]);
+        odata[n][c][h2][w2] = static_cast<scalar_t>(val);
```
thanks, the variable rename from data2 is appreciated :)

ezyang commented on aten/src/ATen/native/cuda/UpSampleBilinear2d.cu:
```diff
@@ -1,25 +1,312 @@
+
+template <typename scalar_t, typename accscalar_t>
+#ifdef __HIP_PLATFORM_HCC__
+C10_LAUNCH_BOUNDS_1(1024)
```
To be applied unconditionally, after fixing kernel launch site!

xmnlab replied:
@ezyang I have applied the changes pointed in this PR, just not sure now what approach I should take to fix the kernel launch site. do you have any recommendation here?

ezyang replied:
At a high level, what you need to do is make sure you never launch the kernel with bounds greater than the bounds you hard-code here (1024). More concretely, you should change the launch code to stop pulling out the limit from the configuration, and just configure the bounds themselves. I'd recommend reading some references about how to do cuda launches. If you're still confused, shout and I can figure it out (this is also my first time working with this code.)

ngimel replied:
The limit still has to be pulled from config, and then a min taken on it (there's no guarantee some future hardware won't have a limit of 512 instead of 1024).

xmnlab replied:
well I've investigated more about this topic, but I didn't find a good answer. 
I just changed the maxthread to `at::cuda::getCurrentDeviceProperties()->maxThreadsPerBlock / 2` and locally it passed. if you have any other idea to address that or any reference to read, that would be very helpful :)

ezyang replied:
Sorry about not responding to this earlier.  I don't think dividing my two is the correct thing to do here; it doesn't seem justified by the semantics of the parameters involved.

What Natalia wrote in her comment is exactly what should be done. The max thread should be `std::min(at::cuda::getCurrentDeviceProperties()->maxThreadsPerBlock, 1024)`, where `1024` is the number that was given to the `LAUNCH_BOUNDS` annotation. This prevents us attempting to launch the kernel with more threads than the device supports (the first argument), or 1024 (the second argument.)

Your change probably results in a launch bounds which is less than either of these limits, but it probably reduces the parallelism that is possible here.

xmnlab replied:
oh I see ... I understand it now ... thanks for the explanation!

ezyang commented on aten/src/ATen/native/cuda/UpSampleBilinear2d.cu:
```diff
@@ -1,25 +1,312 @@
+// Backward (adjoint) operation 1 <- 2 (accumulates)
+template <typename scalar_t, typename accscalar_t>
+#ifdef __HIP_PLATFORM_HCC__
+C10_LAUNCH_BOUNDS_1(1024)
```
To be applied unconditionally, after fixing kernel launch site!

ezyang commented:
Just as a clarification: I'm OK with having changes which are improvements over the old kernels come in latter patches.

ezyang commented on aten/src/ATen/native/cuda/UpSampleBilinear2d.cu:
```diff
@@ -1,25 +1,312 @@
+            input_width, output_width, align_corners);
+
+        upsample_bilinear2d_backward_out_frame<scalar_t, accscalar_t>
+            <<<(num_kernels + num_threads - 1) / num_threads,
```
Arguably we should just add a `ceil_div` function and use it, it's a lot clearer. Doesn't have to be this PR.

xmnlab replied:
sounds good. do you have any idea what file this function should be?

ezyang replied:
Not really. I guess we could always dump it in `Utils.h` for a start, perhaps in the at::impl namespace so that users don't start using it.

xmnlab replied:
sounds good! :) thanks for the feedback!

ngimel replied:
https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/cuda/CUDAApplyUtils.cuh#L615-L618

xmnlab replied:
awesome! thanks @ngimel I will change to that!

ezyang commented:
I finished reviewing one of the kernels, and did a cursory scan of the others. In general the patch looks like it's in good shape. I'd suggest that we fix up the major correctness issues (PackedTensorAccessor ref, and int32 versus int64 indexing) and anything small that is easy to fix (for example, double zeroing the array), and leave other improvements for follow ups. Should be ready to land soon!

xmnlab commented:
thanks so much @ezyang!
I will let you know when it is ready again for a new review! thanks!

xmnlab commented on aten/src/ATen/native/cuda/UpSampleBicubic2d.cu:
```diff
@@ -1,25 +1,335 @@
+    for (int n = 0; n < batchsize; n++) {
+      for (int c = 0; c < channels; ++c) {
+        const scalar_t val = odata[n][c][output_y][output_x];
+        idata[n][c][output_y][output_x] += val;
```
makes sense to me. I am doing that thanks!

xmnlab commented on aten/src/ATen/native/cuda/UpSample.cuh:
```diff
@@ -0,0 +1,254 @@
+  return a > b ? a : b;
+}
+
+static inline void upsample_1d_shape_check(
```
I think it maybe will be done in a future ... if I understood correctly there is a plan to combine functions to work for N dimensions.

xmnlab commented:
@skrah @ezyang 
not sure but the errors on CI seems to be related to jenkins ... it seems all these jobs that failed (for building) ran by 01h 01min ... not sure if it is a coincidence ...

rgommers commented:
@xmnlab indeed it looks like all jobs were aborted at the same time. no obvious issues in the build log related to your code. Comparing e.g. the `cuda9-cudnn7` build with a successful one from another PR, it takes 1hr 14min there and your build is about at the place where the other one is after an hour.

I suggest to just push a new commit to rebuild. Probably a temporary CI hiccup.

ezyang commented:
I accidentally rebooted Jenkins yesterday which is the likely cause, my apologies.

@pytorchbot retest this please

xmnlab commented:
@ezyang @skrah @rgommers 

all tests passed except `pr/caffe2-py2-cuda9.0-cudnn7-windows-build`:
 
```
14:43:49 Build timed out (after 180 minutes). Marking the build as failed.
14:43:49 Build was aborted
14:43:49 [BFA] Scanning build for known causes...
14:43:49 [BFA] No failure causes found
14:43:49 [BFA] Done. 0s
14:43:49 Finished: FAILURE
```

not sure if this timeout means that the code now is slower.

what do you think? do you have any suggestion?

ezyang commented:
Sometimes the Windows build flakes out like that. It didn't timeout while running a relevant test, so I judge it to be not your problem.

ezyang commented:
The launch bounds logic is wrong, but I acknowledge that this is a big patch already; just fix it in a follow up. I am going to go ahead and land this.

xmnlab commented:
sounds good @ezyang thanks!
